src/base/utils/foreignapps.cpp

```diff
@@ -254,20 +258,36 @@ bool Utils::ForeignApps::PythonInfo::isSupportedVersion() const
 PythonInfo Utils::ForeignApps::pythonInfo()
 {
     static PythonInfo pyInfo;
-    if (!pyInfo.isValid())
+
+    if (const QString preferredPythonPath = Preferences::instance()->getPythonExecutablePath().toString();
+        !preferredPythonPath.isEmpty() && (preferredPythonPath != pyInfo.executableName))
+    {
+        if (testPythonInstallation(preferredPythonPath, pyInfo))
+            return pyInfo;
+        LogMsg(QCoreApplication::translate("Utils::ForeignApps", "Failed to find Python executable. Path: \"%1\".")
+            .arg(preferredPythonPath), Log::WARNING);
+    }
+    else if (!pyInfo.isValid())
     {
+        // auto detect only when there are no preferred python path
+
         if (testPythonInstallation(u"python3"_s, pyInfo))
             return pyInfo;
+        LogMsg(QCoreApplication::translate("Utils::ForeignApps", "Failed to find `python3` executable in PATH environment variable. PATH: \"%1\"")
+            .arg(qEnvironmentVariable("PATH")), Log::INFO);
 
         if (testPythonInstallation(u"python"_s, pyInfo))
             return pyInfo;
+        LogMsg(QCoreApplication::translate("Utils::ForeignApps", "Failed to find `python` executable in PATH environment variable. PATH: \"%1\"")
+            .arg(qEnvironmentVariable("PATH")), Log::INFO);
 
 #if defined(Q_OS_WIN)
         if (testPythonInstallation(findPythonPath(), pyInfo))
             return pyInfo;
+        LogMsg(QCoreApplication::translate("Utils::ForeignApps", "Failed to find `python` executable in Windows Registry."), Log::INFO);
 #endif
 
-        LogMsg(QCoreApplication::translate("Utils::ForeignApps", "Python not detected"), Log::INFO);
+        LogMsg(QCoreApplication::translate("Utils::ForeignApps", "Failed to find Python executable"), Log::WARNING);
```

Comment:
I would expect to have the following prerequisite:
```c++
const QString preferredPythonPath = Preferences::instance()->getPythonExecutablePath().toString();
if (pyInfo.isValid() && (preferredPythonPath == pyInfo.executableName))
    return pyInfo;
```
...and then:
```c++
if (!preferredPythonPath.isEmpty())
    // process custom python path
else
    // try to auto detect python
```

Reply:
Otherwise the function has inconsistent behavior.

Reply:
Done.